### ISSUE TITLE
Unknown type varchar error on Nullable columns

### ISSUE THREAD

**Issue by durchgedreht** [xzkostyan/clickhouse-sqlalchemy#85]
I'm using native driver with clickhouse 20.1.4.14 and latest clickhouse-sqlalchemy from PIP.

It seems the issue was coming with clickhouse update (we were using 19.1.x before), but this is a guess as we switched to 20.1 changed driver from sqlalchemy-clickhouse to clickhouse-sqlalchemy due to some other issues. 

So when running a very basic selection of a varchar field, we get an error:

`select description from my_database.creatives; `

The corresponding error is attached below with tracelog. Interestingly I figured the toString() function will make the query work:

`select toString(description) from my_database.creatives; `

Somehow it might be related to Nullable columns, as this column is defined as follows. Note the fields of the query are not null in this case, so there is real data coming back!

```
CREATE TABLE IF NOT EXISTS `my_database`.`creatives` (
  `id` Int64,
  `description` Nullable(varchar),
  ...
)
ENGINE = MergeTree
PRIMARY KEY (`id`)
ORDER BY id;
``` 

I'd be happy about any feedback. Thanks! 
PS: Sorry for the bad formatting of the trace

```2020-02-19 17:49:44,068:DEBUG:clickhouse_driver.connection:Connected to ClickHouse server version 20.1.4, revision: 54431
2020-02-19 17:49:44,069:DEBUG:clickhouse_driver.connection:Query: select version()
2020-02-19 17:49:44,069:DEBUG:clickhouse_driver.connection:Block send time: 0.000063
2020-02-19 17:49:44,071:INFO:root:Running statement 1 out of 1
2020-02-19 17:49:44,083:INFO:root:Parsing with sqlparse statement select description from my_database.creatives
2020-02-19 17:49:44,085:INFO:root:Parsing with sqlparse statement select description from my_database.creatives
2020-02-19 17:49:44,086:INFO:root:Running query: 
select description from spearad_data.creatives
LIMIT 1000
2020-02-19 17:49:44,086:DEBUG:clickhouse_driver.connection:Connecting. Database: my_database. User: default
2020-02-19 17:49:44,089:DEBUG:clickhouse_driver.connection:Connected to ClickHouse server version 20.1.4, revision: 54431
2020-02-19 17:49:44,089:DEBUG:clickhouse_driver.connection:Query: select description from my_database.creatives
LIMIT 1000
2020-02-19 17:49:44,089:DEBUG:clickhouse_driver.connection:Block send time: 0.000060
2020-02-19 17:49:44,091:DEBUG:root:[36m[stats_logger] (timing) sqllab.query.time_executing_query | 4.578125 [0m
2020-02-19 17:49:44,091:ERROR:root:Orig exception: Code: 50. Unknown type varchar
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/columns/service.py", line 74, in get_column_by_spec
    cls = column_by_type[spec]
KeyError: 'varchar'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/clickhouse_sqlalchemy/drivers/native/connector.py", line 156, in execute
    **execute_kwargs
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/client.py", line 214, in execute
    columnar=columnar
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/client.py", line 337, in process_ordinary_query
    columnar=columnar)
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/client.py", line 83, in receive_result
    return result.get_result()
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/result.py", line 48, in get_result
    for packet in self.packet_generator:
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/client.py", line 95, in packet_generator
    packet = self.receive_packet()
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/client.py", line 109, in receive_packet
    packet = self.connection.receive_packet()
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/connection.py", line 386, in receive_packet
    packet.block = self.receive_data()
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/connection.py", line 449, in receive_data
    return self.block_in.read()
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/streams/native.py", line 75, in read
    self.fin)
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/columns/service.py", line 83, in read_column
    column = get_column_by_spec(column_spec, column_options=column_options)
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/columns/service.py", line 63, in get_column_by_spec
    return create_nullable_column(spec, create_column_with_options)
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/columns/nullablecolumn.py", line 5, in create_nullable_column
    nested = column_by_spec_getter(inner)
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/columns/service.py", line 45, in create_column_with_options
    return get_column_by_spec(x, column_options)
  File "/usr/local/lib/python3.6/site-packages/clickhouse_driver/columns/service.py", line 78, in get_column_by_spec
    raise errors.UnknownTypeError('Unknown type 
{}
'.format(e.args[0]))
clickhouse_driver.errors.UnknownTypeError: Code: 50. Unknown type varchar
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/superset/superset/sql_lab.py", line 204, in execute_sql_statement
    db_engine_spec.execute(cursor, sql, async_=True)
  File "/home/superset/superset/db_engine_specs/base.py", line 493, in execute
    cursor.execute(query)
  File "/usr/local/lib/python3.6/site-packages/clickhouse_sqlalchemy/drivers/native/connector.py", line 160, in execute
    raise DatabaseException(orig)
clickhouse_sqlalchemy.exceptions.DatabaseException: Orig exception: Code: 50. Unknown type varchar
10.192.10.21 - - [19/Feb/2020 17:49:44] "POST /superset/sql_json/ HTTP/1.1" 500 -
2020-02-19 17:49:44,111:INFO:werkzeug:10.192.10.21 - - [19/Feb/2020 17:49:44] "POST /superset/sql_json/ HTTP/1.1" 500 -
2020-02-19 17:49:44,255:DEBUG:root:[36m[stats_logger] (incr) queries[0m
10.192.10.21 - - [19/Feb/2020 17:49:44] "GET /superset/queries/1582128376000 HTTP/1.1" 200 -
2020-02-19 17:49:44,263:INFO:werkzeug:10.192.10.21 - - [19/Feb/2020 17:49:44] "GET /superset/queries/1582128376000 HTTP/1.1" 200 -
127.0.0.1 - - [19/Feb/2020 17:49:46] "GET /health HTTP/1.1" 200 -
2020-02-19 17:49:46,382:INFO:werkzeug:127.0.0.1 - - [19/Feb/2020 17:49:46] "GET /health HTTP/1.1" 200 -
10.192.11.13 - - [19/Feb/2020 17:50:03] "GET / HTTP/1.1" 302 -
2020-02-19 17:50:03,099:INFO:werkzeug:10.192.11.13 - - [19/Feb/2020 17:50:03] "GET / HTTP/1.1" 302 -
10.192.10.21 - - [19/Feb/2020 17:50:03] "GET / HTTP/1.1" 302 -
2020-02-19 17:50:03,223:INFO:werkzeug:10.192.10.21 - - [19/Feb/2020 17:50:03] "GET / HTTP/1.1" 302 -
127.0.0.1 - - [19/Feb/2020 17:50:16] "GET /health HTTP/1.1" 200 -
2020-02-19 17:50:16,818:INFO:werkzeug:127.0.0.1 - - [19/Feb/2020 17:50:16] "GET /health HTTP/1.1" 200 -
10.192.11.13 - - [19/Feb/2020 17:50:33] "GET / HTTP/1.1" 302 -
2020-02-19 17:50:33,104:INFO:werkzeug:10.192.11.13 - - [19/Feb/2020 17:50:33] "GET / HTTP/1.1" 302 -
10.192.10.21 - - [19/Feb/2020 17:50:33] "GET / HTTP/1.1" 302 -
2020-02-19 17:50:33,230:INFO:werkzeug:10.192.10.21 - - [19/Feb/2020 17:50:33] "GET / HTTP/1.1" 302 -
127.0.0.1 - - [19/Feb/2020 17:50:47] "GET /health HTTP/1.1" 200 -
2020-02-19 17:50:47,249:INFO:werkzeug:127.0.0.1 - - [19/Feb/2020 17:50:47] "GET /health HTTP/1.1" 200 -
10.192.11.13 - - [19/Feb/2020 17:51:03] "GET / HTTP/1.1" 302 -```

**Comment by xzkostyan** [comment]
There was pull request in 20 version of ClickHouse server that saves type synonyms in table declarations and in data formats. VARCHAR is synonym for String. Earlier server versions have types canonization. So even you specify VARCHAR as source type you will get String as canonized type.

Many 3rd party tools are broken now. ClickHouse developers say that that PR is contentious and can be reverted or can lead to another changes.

I'll prefer to wait for server developers' final decision on types canonization.

**Comment by durchgedreht** [comment]
**Many 3rd party tools are broken now.**
You name it! Thanks so much for the explanation!

Would anyone be able to point me to the pull request as I have hard times finding it and would like to understand the behaviour in full?

I can still proactively decide how to roll forward and would like to make sure I take the right decision what to use where (table DDL and creation / driver / tool). Not all are 100% under my control, so this might get a bit tricky. Currently using VARCHAR in the DDL definition using JDBC to create table, sqlalchemy (native) to read (VARCHAR) which only works when calling toString() in the statement itself. Looks quite ugly and I'd like to streamline this mess! Would using String clean this up or will this cause problems when the Pull request is reverted? Unfortunately I cannot wait that much longer and take a decision now.

This is really annoying, especially as this went into the release and there's nothing mentioned in the backwards incompatibilities list (and other fixes needed to go into production):
https://github.com/ClickHouse/ClickHouse/blob/master/CHANGELOG.md

**Comment by xzkostyan** [comment]
I've found PR: https://github.com/ClickHouse/ClickHouse/pull/8370

**Comment by durchgedreht** [comment]
Thanks very much, this is very useful! Highly appreciate your update. Will close issue here as this is more Clickhouse dependant than a driver issue!

**Comment by xzkostyan** [comment]
I don't know on which side this should be fixed, but it seems we can make workaround in driver.

**Comment by xzkostyan** [comment]
PR with discussion about canonical types: https://github.com/ClickHouse/ClickHouse/pull/8429#issuecomment-589719114

**Comment by xzkostyan** [comment]
Keep alias type in metadata [was reverted](https://github.com/ClickHouse/ClickHouse/pull/9594). This fix is present in 20.3 version of server.
